Mathlib/GroupTheory/GroupAction/Hom.lean

```diff
@@ -1,212 +1,524 @@
+attribute [simp] CompTripleClass.comp_eq
+
+namespace CompTripleClass
+
+lemma comp_id (F : Type _) {N P : outParam (Type _)} (ψ : outParam (N → P)) :
+    CompTripleClass F (@id N) ψ ψ := {comp_eq := rfl}
+
+lemma id_comp (F : Type _) {M N : outParam (Type _)} (φ : outParam (M → N)) :
+    CompTripleClass F φ (@id N) φ := {comp_eq := rfl}
+
+lemma comp (F : Type _) {M N P : outParam (Type _)}
+    (φ : outParam (M → N)) (ψ : outParam (N → P)) :
+    CompTripleClass F φ ψ (ψ.comp φ) := {comp_eq := rfl}
+
+lemma comp_inv {M N : outParam (Type _)} {φ : outParam (M → N)} {ψ :outParam (N → M)}
+    (h : Function.RightInverse ↑φ ↑ψ) :
+    CompTriple φ ψ id := {comp_eq := h.id }
+
+lemma apply {M N P : outParam (Type _)}
+    {φ : outParam (M → N)} {ψ : outParam (N → P)} {χ : outParam (M → P)}
+    (h : CompTriple φ ψ χ) (x : M) :
+    ψ (φ x) = χ x := by
+  rw [← h.comp_eq, Function.comp_apply]
+
+@[simp]
+theorem comp_assoc {M N P Q : outParam (Type _)}
+    {φ₁ : outParam (M → N)} {φ₂ : outParam (N → P)} {φ₁₂ : outParam (M → P)}
+    (κ : CompTriple φ₁ φ₂ φ₁₂)
+    {φ₃ : outParam (P → Q)} {φ₂₃ : outParam (N → Q)} (κ' : CompTriple φ₂ φ₃ φ₂₃)
+    {φ₁₂₃ : outParam (M → Q)} :
+    CompTriple φ₁ φ₂₃ φ₁₂₃ ↔ CompTriple φ₁₂ φ₃ φ₁₂₃ := by
+  constructor
+  all_goals {
+    rintro ⟨h⟩
+    exact ⟨ by simp only [← κ.comp_eq, ← h, ← κ'.comp_eq]; rfl ⟩ }
+
+end CompTripleClass
+
+end CompTripleClass
+
+-/
 
 assert_not_exists Submonoid
 
-variable (M' : Type*)
-variable (X : Type*) [SMul M' X]
-variable (Y : Type*) [SMul M' Y]
-variable (Z : Type*) [SMul M' Z]
-variable (M : Type*) [Monoid M]
-variable (A : Type*) [AddMonoid A] [DistribMulAction M A]
-variable (A' : Type*) [AddGroup A'] [DistribMulAction M A']
-variable (B : Type*) [AddMonoid B] [DistribMulAction M B]
-variable (B' : Type*) [AddGroup B'] [DistribMulAction M B']
-variable (C : Type*) [AddMonoid C] [DistribMulAction M C]
-variable (R : Type*) [Semiring R] [MulSemiringAction M R]
-variable (R' : Type*) [Ring R'] [MulSemiringAction M R']
-variable (S : Type*) [Semiring S] [MulSemiringAction M S]
-variable (S' : Type*) [Ring S'] [MulSemiringAction M S']
-variable (T : Type*) [Semiring T] [MulSemiringAction M T]
+section MulActionHom
+
+variable {M' : Type*}
+variable {M : Type*} {N : Type*} {P : Type*}
+variable (φ : M → N) (ψ : N → P) (χ : M → P)
+variable (X : Type*) [SMul M X] [SMul M' X]
+variable (Y : Type*) [SMul N Y] [SMul M' Y]
+variable (Z : Type*) [SMul P Z]
 
 /-- Equivariant functions. -/
 -- Porting note: This linter does not exist yet
 -- @[nolint has_nonempty_instance]
 structure MulActionHom where
   /-- The underlying function. -/
-  protected toFun : X → Y
-  /-- The proposition that the function preserves the action. -/
-  protected map_smul' : ∀ (m : M') (x : X), toFun (m • x) = m • toFun x
-#align mul_action_hom MulActionHom
+  toFun : X → Y
+  /-- The proposition that the function commutes with the actions. -/
+  map_smul' : ∀ (m : M) (x : X), toFun (m • x) = (φ m) • toFun x
```

Comment:
Is there a reason to remove the `protected` here? I see quite a few instances of needing to insert the `_root_` namespace below and I prefer over-namespacing.

Reply:
No, there is not; this is not an issue I was enough aware of.